numpyro/distributions/continuous.py

```diff
@@ -290,6 +291,118 @@ def infer_shapes(concentration):
         return batch_shape, event_shape
 
 
+class EulerMaruyama(Distribution):
+    arg_constraints = {"t": constraints.ordered_vector}
+
+    def __init__(self, t, sde_fn, init_dist, validate_args=None):
+        self.t = t
+        self.dt = jnp.diff(t)
+        self.sde_fn = sde_fn
+        self.init_dist = init_dist
+
+        if init_dist.batch_shape:
+            raise ValueError(
+                "Init distribution is expected to have no batch dimension."
+            )
+        if jnp.ndim(t) != 1:
+            raise ValueError("`t` is expected to have 1 dimension.")
+        event_shape = t.shape + init_dist.event_shape
+        batch_shape = ()
+
+        super(EulerMaruyama, self).__init__(
+            batch_shape, event_shape, validate_args=validate_args
+        )
+
+    @constraints.dependent_property
+    def support(self):
+        return constraints.independent(constraints.real, self.event_dim)
+
+    def sample(self, key, sample_shape=()):
+        assert is_prng_key(key)
+        sample_size = np.prod(sample_shape)
+        sample_shape_ = (sample_size,) if sample_shape else ()
+
+        def step(y_curr, xs):
+            noise_curr, t_curr, dt_curr = xs
+            f, g = self.sde_fn(y_curr, t_curr)
+            mu = y_curr + dt_curr * f
+            sigma = jnp.sqrt(dt_curr) * g
+            y_next = mu + sigma * noise_curr
+            return y_next, y_next
+
+        noises = random.normal(
+            key, shape=sample_shape_ + (self.event_shape[0] - 1,) + self.event_shape[1:]
+        )
+        inits = self.init_dist.sample(key, sample_shape=sample_shape_)
+
+        def scan_fn(init, noise):
+            return scan(step, init, (noise, self.t[:-1], self.dt))
+
+        if sample_shape:
+            _, sde_out = vmap(scan_fn)(inits, noises)
+            sde_out = jnp.concatenate([inits[:, None], sde_out], axis=1)
+            sde_out = jnp.reshape(sde_out, sample_shape + self.event_shape)
+        else:
+            _, sde_out = scan_fn(inits, noises)
+            sde_out = jnp.concatenate([inits[None], sde_out], axis=0)
+
+        return sde_out
+
+    @validate_sample
+    def log_prob(self, value):
+        sample_shape = value.shape[: -self.event_dim]
+
+        if sample_shape:
+            reshaped_value = value.reshape((-1,) + self.event_shape)
+            xtm1 = reshaped_value[:, :-1]
+            xt = reshaped_value[:, 1:]
+            value0 = reshaped_value[:, 0]
+
+            f, g = vmap(vmap(self.sde_fn), (0, None))(xtm1, self.t[:-1])
+            dt = self.dt.reshape(self.dt.shape + (1,) * (self.event_dim - 1))
+
+            if len(f.shape) - 1 < len(dt.shape):
+                f = f.reshape((reshaped_value.shape[0],) + dt.shape)
+            if len(g.shape) - 1 < len(dt.shape):
```

Comment:
> I guess sde_fn might return scalar, rather than event_shape?

I assume as you say.
I made test case referring to pymc [example](https://docs.pymc.io/en/v3/pymc-examples/examples/time_series/Euler-Maruyama_and_SDEs.html).

> `batch_dim = 2 if sample_shape else 1`

Why isn't it `batch_dim = 1 if sample_shape else 0` ?

Reply:
You can specify batch_dim = 1 if sample_shape else 0. Then we need to add additional event dimensions for f and g as follows
```
f = f.reshape(f.shape[:batch_dim+1] + (1,) * (xt.ndim - f.ndim) + f.shape[batch_dim+1:])
```
In above comment, batch_dim includes time dimension. In this comment, batch_dim does not include time dimension. This statement adds missing event dimensions on the right of time dimension.

Reply:
Thank you. I understand how this work.
I used your code.